File: utils/appConfig.ts
/**
 * Application origin (`BASE_URL`) resolution from environment.
 */

function normalizeBaseUrlCandidate(raw: string): string | null {
  const normalized = raw.trim().replace(/^["']|["']$/g, '').replace(/\/+$/, '');
  if (
    !normalized ||
    normalized.toLowerCase() === 'undefined' ||
    normalized.toLowerCase() === 'null' ||
    !/^https?:\/\//i.test(normalized)
  ) {
    return null;
  }
  return normalized;
}

/** Resolve app origin from an optional override, then required `BASE_URL` env. */
export function resolveBaseUrl(raw?: string): string {
  if (raw != null) {
    const fromArg = normalizeBaseUrlCandidate(raw);
    if (fromArg) return fromArg;
  }

  const fromEnv = normalizeBaseUrlCandidate(process.env.BASE_URL ?? '');
  if (fromEnv) return fromEnv;

  throw new Error(
    'BASE_URL is required. Set the BASE_URL environment variable in `.env`.'
  );
}

/** Origin only (scheme + host + port), for host-matching helpers. */
export function resolveAppOrigin(raw?: string): string {
  const base = resolveBaseUrl(raw);
  try {
    return new URL(base).origin;
  } catch {
    return base;
  }
}

File: utils/app_config.py
"""Mirror of `utils/appConfig.ts` for Python subprocesses (invoice Excel writer)."""

from __future__ import annotations

import os


def _normalize_base_url_candidate(raw: str) -> str | None:
    normalized = raw.strip().strip("'\"").rstrip("/")
    if not normalized or normalized.lower() in ("undefined", "null"):
        return None
    if not normalized.lower().startswith(("http://", "https://")):
        return None
    return normalized


def resolve_base_url(raw: str | None = None) -> str:
    if raw is not None:
        from_arg = _normalize_base_url_candidate(raw)
        if from_arg:
            return from_arg

    from_env = _normalize_base_url_candidate(os.environ.get("BASE_URL", ""))
    if from_env:
        return from_env

    raise RuntimeError(
        "BASE_URL is required. Set the BASE_URL environment variable in `.env`."
    )